internal/utils/graph_utils.go

```diff
@@ -0,0 +1,41 @@
+package utils
+
+import (
+	"context"
+)
+
+type (
+	EngineUtils interface {
+		CheckVisited(ctx context.Context, current string) (context.Context, bool)
+	}
+	EngineUtilsProvider struct {
+		EngineUtils
+	}
+	contextKey string
+)
+
+const visitedMapKey = contextKey("visitedMap")
+
+func (*EngineUtilsProvider) CheckVisited(ctx context.Context, current string) (context.Context, bool) {
```

Comment:
As this returns a new context with the already visited value, and the result, let's maybe rather call this `CheckAndAddVisited` or something more descriptive.
I would also prefer to rather pass a subject here. You can still use the string representation inside of that, but it will make it less error prone going forward.
```suggestion
func (*EngineUtilsProvider) CheckAndAddVisited(ctx context.Context, current relationtuple.Subject) (context.Context, bool) {
```